diff --git a/llvm/lib/Target/AArch64/AArch64.h b/llvm/lib/Target/AArch64/AArch64.h
index 76f55666e..985f67968 100644
--- a/llvm/lib/Target/AArch64/AArch64.h
+++ b/llvm/lib/Target/AArch64/AArch64.h
@@ -71,6 +71,9 @@ FunctionPass *createAArch64StackTaggingPass(bool IsOptNone);
 FunctionPass *createAArch64StackTaggingPreRAPass();
 ModulePass *createAArch64GlobalsTaggingPass();
 
+// JUPITER ADDED
+FunctionPass *createAArch64MachineInstrPrinterPass();
+
 void initializeAArch64A53Fix835769Pass(PassRegistry&);
 void initializeAArch64A57FPLoadBalancingPass(PassRegistry&);
 void initializeAArch64AdvSIMDScalarPass(PassRegistry&);
@@ -106,6 +109,9 @@ void initializeFalkorMarkStridedAccessesLegacyPass(PassRegistry&);
 void initializeLDTLSCleanupPass(PassRegistry&);
 void initializeSMEABIPass(PassRegistry &);
 void initializeSVEIntrinsicOptsPass(PassRegistry &);
+
+// JUPITER ADDED
+void initializeAArch64MachineInstrPrinterPass(PassRegistry &);
 } // end namespace llvm
 
 #endif
diff --git a/llvm/lib/Target/AArch64/AArch64AddStaticAnalysisInformation.cpp b/llvm/lib/Target/AArch64/AArch64AddStaticAnalysisInformation.cpp
new file mode 100644
index 000000000..642479bbc
--- /dev/null
+++ b/llvm/lib/Target/AArch64/AArch64AddStaticAnalysisInformation.cpp
@@ -0,0 +1,52 @@
+#include "AArch64.h"
+#include "AArch64InstrInfo.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+
+using namespace llvm;
+
+#define AARCH64_MACHINEINSTR_PRINTER_PASS_NAME "Dummy AArch64 machineinstr printer pass"
+
+namespace {
+
+class AArch64MachineInstrPrinter : public MachineFunctionPass {
+public:
+    static char ID;
+
+    AArch64MachineInstrPrinter() : MachineFunctionPass(ID) {
+        initializeAArch64MachineInstrPrinterPass(*PassRegistry::getPassRegistry());
+    }
+
+    bool runOnMachineFunction(MachineFunction &MF) override;
+
+    StringRef getPassName() const override { return AARCH64_MACHINEINSTR_PRINTER_PASS_NAME; }
+};
+
+char AArch64MachineInstrPrinter::ID = 0;
+
+bool AArch64MachineInstrPrinter::runOnMachineFunction(MachineFunction &MF) {
+
+    for (auto &MBB : MF) {
+        outs() << "Contents of MachineBasicBlock:\n";
+        outs() << MBB << "\n";
+        const BasicBlock *BB = MBB.getBasicBlock();
+        outs() << "Contents of BasicBlock corresponding to MachineBasicBlock:\n";
+        outs() << BB << "\n";
+    }
+
+    return false;
+}
+
+} // end of anonymous namespace
+
+INITIALIZE_PASS(AArch64MachineInstrPrinter, "AArch64-machineinstr-printer",
+    AARCH64_MACHINEINSTR_PRINTER_PASS_NAME,
+    true, // is CFG only?
+    true  // is analysis?
+)
+
+namespace llvm {
+
+FunctionPass *createAArch64MachineInstrPrinterPass() { return new AArch64MachineInstrPrinter(); }
+
+}
\ No newline at end of file
diff --git a/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp b/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp
index 559879139..5c7de1277 100644
--- a/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp
+++ b/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp
@@ -239,6 +239,9 @@ extern "C" LLVM_EXTERNAL_VISIBILITY void LLVMInitializeAArch64Target() {
   initializeAArch64LowerHomogeneousPrologEpilogPass(*PR);
   initializeAArch64DAGToDAGISelPass(*PR);
   initializeAArch64GlobalsTaggingPass(*PR);
+
+  // JUPITER ADDED
+  initializeAArch64MachineInstrPrinterPass(*PR);
 }
 
 //===----------------------------------------------------------------------===//
@@ -758,6 +761,9 @@ void AArch64PassConfig::addPreRegAlloc() {
     // be register coalescer friendly.
     addPass(&PeepholeOptimizerID);
   }
+
+  // JUPITER ADDED
+  addPass(createAArch64MachineInstrPrinterPass());
 }
 
 void AArch64PassConfig::addPostRegAlloc() {
diff --git a/llvm/lib/Target/AArch64/CMakeLists.txt b/llvm/lib/Target/AArch64/CMakeLists.txt
index ada3a5730..6e0ba31f9 100644
--- a/llvm/lib/Target/AArch64/CMakeLists.txt
+++ b/llvm/lib/Target/AArch64/CMakeLists.txt
@@ -87,6 +87,8 @@ add_llvm_target(AArch64CodeGen
   SVEIntrinsicOpts.cpp
   AArch64SIMDInstrOpt.cpp
 
+  AArch64AddStaticAnalysisInformation.cpp
+
   DEPENDS
   intrinsics_gen
 
diff --git a/machine-pass.patch b/machine-pass.patch
new file mode 100644
index 000000000..e69de29bb
